probe: Convert read error to failure at http probe

hshiina opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug

#### What this PR does / why we need it:
If `DoHTTPProbe()` returns an error as third return value, a probe worker treats it as a client failure and throws it away:
https://github.com/kubernetes/kubernetes/blob/5973accf487eef985393cde200e7a4f9d09bc2e0/pkg/probe/http/http.go#L102-L108
https://github.com/kubernetes/kubernetes/blob/5973accf487eef985393cde200e7a4f9d09bc2e0/pkg/kubelet/prober/prober.go#L100-L111
https://github.com/kubernetes/kubernetes/blob/5973accf487eef985393cde200e7a4f9d09bc2e0/pkg/kubelet/prober/worker.go#L287-L291

When an error is raised while a prober is reading a response body, something is likely wrong with a server (container) such as hang up or abortion. This change returns `Failure` without an error in order to handle it as a server (container) failure in this case.

This PR also adds test cases for HTTP probe with failure modes such as disconnection, hang up and slow down, which are emulated by `agnhost liveness`.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #122591

#### Special notes for your reviewer:
This test requires the PR  #125489 to add a function to agnhost.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixed a bug that a container is not restarted when an HTTP liveness probe hits an unexpected disconnection or a timeout while receiving a response body.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aojea commented on test/e2e/common/node/container_probe.go:
```diff
@@ -728,6 +728,32 @@ done
+		             The pod is never restarted because the error is ignored (issue #122591).
+
+	*/
+	ginkgo.It("should *not* be restarted with a liveness probe that is disconnected unexpectedly", func(ctx context.Context) {
```
this is the current behavior, right?

hshiina replied:
As far as I tried, this is the current behavior.

In this case, `unexpected EOF` error is returned from `ReadAtMost()`. Then, both  `Failure` result and this error are returned:
https://github.com/kubernetes/kubernetes/blob/fb7bbd2e5df36212ca5133bcbaa46dcdb4d18f0e/pkg/probe/http/http.go#L102-L109

The error is checked prior to the failure. So, this failure is ignored:
https://github.com/kubernetes/kubernetes/blob/fb7bbd2e5df36212ca5133bcbaa46dcdb4d18f0e/pkg/kubelet/prober/worker.go#L287-L291

SergeyKanzhelev commented:
/assign @matthyx

matthyx commented:
@hshiina this is awesome, let's wait for #125489 to run the test
Do you plan to fix the livenessProbe behavior in this PR too?

hshiina commented:
> @hshiina this is awesome, let's wait for #125489 to run the test Do you plan to fix the livenessProbe behavior in this PR too?

Yes, I'm planning to fix it eventually. Because the issue #122591 still has a `needs-information` label, I'm trying to reproduce this issue with the current behavior at first.

hshiina committed:
Convert read error into failure at http probe

If `DoHTTPProbe()` returns an error as third return value, a probe
worker treats it as a client failure and throw it away. When an error is
raised while a prober is reading a response body, something is likely
wrong with a server (container). This change returns `Failure` without
an error in order to handle it as a server (container) failure in this
case.

hshiina committed:
e2e: Add HTTP liveness probe tests for failure modes

hshiina commented:
/hold

SergeyKanzhelev commented on pkg/probe/http/http.go:
```diff
@@ -104,7 +104,8 @@ func DoHTTPProbe(req *http.Request, client GetHTTPInterface) (probe.Result, stri
 		} else {
-			return probe.Failure, "", err
+			// Convert errors into failures to catch timeouts.
+			return probe.Failure, err.Error(), nil
```
Can you please give some description on what this change will result in? More context you can provide, easier it is to review

SergeyKanzhelev commented on pkg/probe/http/http.go:
```diff
@@ -87,24 +87,28 @@ type GetHTTPInterface interface {
 			klog.V(4).Infof("Non fatal body truncation for %s, Response: %v", url.String(), *res)
 		} else {
-			return probe.Failure, "", err
+			// Something seems wrong with the HTTP server (timeouts or disconnection). Return Failure (not error).
```
will this result in a different outcome of the probe check? Is it backward compatible change?

hshiina replied:
It’s not backward compatible. I thought this was a bug fix.

There seems to be other options:
- Keep the current behavior for backward compatibility.
- Check the status code and return a Failure only if the code is abnormal (the case mentioned in #122591).

mkumatag commented on test/utils/image/manifest.go:
```diff
@@ -216,7 +216,7 @@ const (
 func initImageConfigs(list RegistryList) (map[ImageID]Config, map[ImageID]Config) {
 	configs := map[ImageID]Config{}
-	configs[Agnhost] = Config{list.PromoterE2eRegistry, "agnhost", "2.53"}
+	configs[Agnhost] = Config{list.PromoterE2eRegistry, "agnhost", "2.54"}
```
There is no `2.54` version present!
```
% manifest-tool inspect registry.k8s.io/e2e-test-images/agnhost:2.54
ERRO[0000] registry.k8s.io/e2e-test-images/agnhost:2.54: not found
ERRO[0000] Unknown descriptor type:
```

hshiina replied:
It's being updated in #125489. That's why I'm holding this PR. I'm sorry for confusing.
